### ISSUE TITLE
2.1.0 is not compatible with netcoreapp1.1/1.0

### ISSUE THREAD

**Issue by Johlandabee** [Mongo2Go/Mongo2Go#35]
I get the following error if I try to add Mongo2Go **2.1.0** to my .Net Core 1.1 project.  
**2.0.0-alpha1** works though.

> Errors in c:\Projects\\****\sln\Test\Test.csproj
> Package Mongo2Go 2.1.0 is not compatible with netcoreapp1.1 (.NETCoreApp,Version=v1.1). Package Mongo2Go 2.1.0 supports: net (.NETFramework,Version=v0.0)
> One or more packages are incompatible with .NETCoreApp,Version=v1.1.

I am using the latest [SDK](https://github.com/dotnet/core/blob/master/release-notes/download-archives/1.0.4-download.md).

**Comment by Johlandabee** [comment]
I guess I found the issue. `dotnet pack` is not used to create the package and the .nuspec contained is missing dependency and target information.

```xml
<?xml version="1.0" encoding="utf-8"?>
<package xmlns="http://schemas.microsoft.com/packaging/2011/08/nuspec.xsd">
  <metadata>
    <id>Mongo2Go</id>
    <version>2.1.0</version>
    <title>Mongo2Go</title>
    <authors>Johannes Hoppe and many contributors</authors>
    <owners>Johannes Hoppe</owners>
    <requireLicenseAcceptance>false</requireLicenseAcceptance>
    <projectUrl>http://github.com/Mongo2Go/Mongo2Go</projectUrl>
    *snip*
    <language>en-US</language>
    <tags>MongoDB Mongo unit test integration runner</tags>
  </metadata>
</package>
```
**vs**

```xml
<?xml version="1.0" encoding="utf-8"?>
<package xmlns="http://schemas.microsoft.com/packaging/2013/05/nuspec.xsd">
  <metadata>
    <id>Mongo2Go</id>
    <version>2.1.0</version>
    <authors>Mongo2Go</authors>
    <owners>Mongo2Go</owners>
    <requireLicenseAcceptance>false</requireLicenseAcceptance>
    <description>Package Description</description>
    <dependencies>
      <group targetFramework=".NETStandard1.6">
        <dependency id="System.Threading.Thread" version="4.3.0" exclude="Build,Analyzers" />
        <dependency id="system.net.networkinformation" version="4.3.0" exclude="Build,Analyzers" />
        <dependency id="NETStandard.Library" version="1.6.1" exclude="Build,Analyzers" />
        <dependency id="System.Diagnostics.Process" version="4.3.0" exclude="Build,Analyzers" />
      </group>
    </dependencies>
  </metadata>
</package>
```

**Comment by twgraham** [comment]
Any update when this will be fixed? I still cant seem to use 2.1.0 in a netcoreapp1.1 project

**Comment by JohannesHoppe** [comment]
Pull requests welcome!
I don't have capacity... :-(

**Comment by twgraham** [comment]
Nw, ill see what i can do

**Comment by Johlandabee** [comment]
@twgraham I'm currently working on my own implementation which targets .net 4.6 and netstandard1.6. Guess I will publish 1.0.0 on nuget this week. If you want, you can check it out: https://github.com/Johlandabee/Amongst

**Comment by JohannesHoppe** [comment]
Out of curiosity, why are you starting a 100% identical project?

**Comment by Johlandabee** [comment]
Well, I actually tried to properly port Mongo2Go but was not happy with the result. Sometimes it is more work to fix a particular problem than starting from scratch. Furthermore, I found a few design decisions I did not like personally. Also I thought this project was kind of abandoned at the moment.

**Comment by JohannesHoppe** [comment]
Sure, I don't have a project with MongoDB right now and I'm writing on a book...
Why don't you join `Mongo2Go`? This project had a lot of maintainers over the years! 😄 👍

**Comment by Johlandabee** [comment]
Alright, that's an offer I can't deny. Then again, how should we continue?

**Comment by JohannesHoppe** [comment]
You have right access! Pretty much everything is possible! ;-)

Right now the biggest pain point is the .NET Core build, which worked on my machine and (maybe) nowhere else?

Currently we have the `Johannes is doing the human-CI` gated-check in. It compiles, I upload to nuget.org. @G3N7 started on some first automation (see https://github.com/Mongo2Go/Mongo2Go/issues/28)

**Comment by JohannesHoppe** [comment]
Also I would be interested, which design decisions you don't like. Maybe we should change that.

**Comment by Johlandabee** [comment]
The question is what to do with [Amongst](https://github.com/Johlandabee/Amongst). It is actually ready, only import/export functionality is missing and I got that working locally. Maybe you can take a look at it and decide if it's worth taking the code and start working from there?

**Comment by Johlandabee** [comment]
My biggest issue right now is the complexity. There are many wrappers that I think, are not necessary. I always try to keep things simple and maintainable, write as little as possible code to archive my goal.

**Comment by JohannesHoppe** [comment]
I see your points. Mongo2Go got a bit complex over time. Also it was 2012 when I started this 👴! A big spring-cleaning would be nice.
My biggest concern is to avoid unnecessary breaking changes. Let's face it: Most people will set up their Testing-Environment once and then just want it to work over the years. Most of the Mongo2Go users are still on .NET 4.6, too! The average feature-request is an update to the MongoDB binaries. 😄 

We could
a) Remove the complexity from Mongo2Go, make it working on all .NET flavors again.
b) Take your Amongst as a starting point for Mongo2Go v3. Change Namespaces and public methods to match the old API. As far as I see, a lot of Amongst was inspired by Mongo2Go anyway.

**Comment by Johlandabee** [comment]
I'd roll with option b. :sweat_smile:
Most of the work is already done.

As for braking changes, I think with a major version jump you are allowed to do so. People should be able to change two lines of their unit tests, especially since Mongo2Go provides only a handful of methods. :smiling_imp:   
But that is only me thinking of a perfect world.

Also here are a few additional points I have:

1) Huston, [we need a license](https://choosealicense.com/no-license/).

2) I think we do not need to cleanup anything after running. CIs do this for us and we have another posibility to debug.
While working locally it shouldn't be a problem do delete the instances folder once in a while.
Amongst solves this with a uniqe data folder for each instance.

3) Why do we search recursively for our mongodb binaries with a pattern? That seems to be written with multiple MongoDB versions in mind. I Thought about that too, but providing more than the current stable build would rediculusly increasy the package size. A solution could be providing multiple versions plugin like via NuGet.

4) We really should not use [natives](https://github.com/Mongo2Go/Mongo2Go/blob/master/src/Mongo2Go/Helper/NativeMethods.cs), why do we set the window title in the first place? 

5) We have two PortWatcher implementations even though the TCPListener solution works perfectly well on windows too.

6) Is pure local debugging really a use case? At least I have mongod installed on my machine for that matter.

**Comment by JohannesHoppe** [comment]
1. I go for MIT
2. open for discussion. maybe just a cleanup for older folders on next run?
3. this was born because of troubles. different locations for the packages folder were possible. also because of the tests of Mongo2Go. here the folder is different as for the final packed version.
4. obsolete code, can be deleted now
5. i agree UnixPortWatcher / TcpListener can do the job for both platforms. `GetActiveTcpListeners` is the older implementation.
6. Definitively required for "Enterprise environments". Helps big time to avoid evil company rules. (no downloads allowed, only after a big process, blah, blah) But Nuget packages are usually allowed and are NOT blocked by evil blacklists.  Also everybody in the team has automatically the same version.

**Comment by Johlandabee** [comment]
1) MIT should be good.
2) Cleanup before the next run could be quite useful i guess.
6) Didn't think about it that way, it's a blessing I do not work for such a company. Definitely a good point. Also I imagine working with Mongo2Go in combination with LinqPad could be quite nice.

As for 2 and 6 we could have something like this to reduce the API:

```csharp
Mongo2Go.MongoDBInstance.Spawn(new MongoDBInstanceOptions {
    LogVerbosity = LogVerbosity.Verbose,
    OutputHelper = output,
    CleanBeforeRun = false,
    SingleInstance = true,
    AllowMultipleRunners = false,
    BinaryDirectory = "/foo/bar"
});
```

**Comment by shawnmclean** [comment]
Any ETA on this?

**Comment by Johlandabee** [comment]
Hi, sorry for the late reply. A version 3.0 of Mongo2Go which solves this issue is currently in the works, but there is **no ETA** since I am pretty busy with work/school at the moment.

**Comment by JohannesHoppe** [comment]
Same here. 3.0 of Mongo2Go will be a complete rewrite, based on the work of @Johlandabee (Amongst) plus the missing "Local debugging" mode. No ETA available.

Meanwhile, we would accept any PR that fixes compatibility with .Net Core 1.1!

**Comment by JohannesHoppe** [comment]
FYI

A new version was pushed by @avrum:
Mongo2Go 2.2.0 contains binaries from MongoDB v3.4.7
(the 1.x branch is still on MongoDB v3.2.7)

**Comment by JohannesHoppe** [comment]
Nobody complained,  so I think it should work again for everybody (on windows). Linux comes back via #42 / #44. I can't test Mac right now (no hardware).
